### PR TITLE
Allow event listener options to be specified with a function in addition to an object or boolean

### LINKED ISSUE #468: [ShadyDOM] Event listener options cannot be specified with a function in IE
### Description

ShadyDOM's `addEventListener` [implementation](https://github.com/webcomponents/polyfills/blob/master/packages/shadydom/src/patch-events.js#L45-L64) only allows non-boolean event listener options passed in as objects. If a function is passed, the event will always capture, which is incorrect. Instead of checking if the options are an object, we can probably check that they are not a boolean.

Note, `noPatch` users cannot rely on this patch anyway and need a separate polyfill if they want to use event options.

### Browsers affected

- [ ] Chrome
- [ ] Firefox
- [ ] Edge
- [ ] Safari
- [X] IE 11

### PR DESCRIPTION
**Pull request by sorvell**
Fixes #468 and #471

* Allows event listener options to be specified with any options that are `instanceof Object`. This allows functions to be used in addition to objects.
* Also fixes `eventPhase` to properly be `Event.AT_TARGET` on retargeted events.

### CODE REVIEW

**Review by bicknellr — APPROVED** [review]
LGTM

**Comment by bicknellr** [comment]
Really odd edge case. I guess the code where you found this must have been _intentionally_ using a function in this way, since the problem is that `capture` isn't set properly in this case?

**Review by kevinpschaaf — APPROVED** [review]
LGTM

File: packages/shadydom/src/patch-events.js [hunk 1]
```diff
@@ -44,7 +44,7 @@ const supportsEventOptions = (() => {
 
 const parseEventOptions = (optionsOrCapture) => {
   let capture, once, passive, shadyTarget;
-  if (optionsOrCapture && typeof optionsOrCapture === 'object') {
+  if (optionsOrCapture && optionsOrCapture instanceof Object) {
```
Comment (justinfagnani) [inline packages/shadydom/src/patch-events.js:47]: I would do:

```suggestion
  if (optionsOrCapture && (typeof optionsOrCapture === 'object' || typeof optionsOrCapture === 'function')) {
```

because `new Boolean(false) instanceof Object` is true, and to keep the same cross-realm behavior. Plus `typeof` is very fast compared to `instanceof`.

**Review by justinfagnani — CHANGES_REQUESTED** [review]

File: packages/shadydom/src/patch-events.js (hunk 1 above)
Comment (sorvell) [reply packages/shadydom/src/patch-events.js:47]: `new Boolean(false) == false` so it wouldn't get past the `&&` here at all.

We can put back the `typeof` as a perf optimization.

Comment (justinfagnani) [reply packages/shadydom/src/patch-events.js:47]: I just mean that any `new {Primitive}()` will return an object, `new Boolean(true)` too. That could conceivably flow into options.

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    origAddEventListener.call(this, type, nativeListener, capture);
+  };
+
+  nativeEventTarget.prototype.removeEventListener = function (
+    type: string,
+    listener: EventListenerOrEventListenerObject | null,
+    options?: boolean | AddEventListenerOptions | undefined
+  ) {
+    const {capture} = parseEventOptions(options);
+    origRemoveEventListener.call(this, type, listener, capture);
```
Comment (bicknellr) [inline packages/webcomponentsjs/ts_src/platform/event-listener-options.ts:83]: This won't remove the correct listener because `listener` here isn't the same listener that's given to `origAddEventListener` above (`nativeListener`). An important thing to note is that keeping a map from the `listener` parameter to the actual listener given to the native function won't be enough to fix this because of how listeners are deduplicated when they're added based on their options.

The 'formdata' event polyfill has to handle a kind of similar situation and has this [`EventListenerArray` class](https://github.com/webcomponents/polyfills/blob/6f17afd6bf42c7383ade9b9819c378c9f504b735/packages/formdata-event/ts_src/event_listener_array.ts) that tracks the deduplication parts. However, the 'formdata' event polyfill only needs to pay attention to listeners for 'formdata' events and isn't actually wrapping these listeners (it only cares when the number of deduplicated 'formdata' event listeners on a target changes between zero and non-zero) so `EventListenerArray` isn't set up to actually provide a full mapping of user-provided listeners to the 'real' listeners; but it would likely be a good starting point for doing that here. There's also a few things in it that aren't needed here: the `push` and `delete` methods are really the important parts.

**Review by bicknellr — CHANGES_REQUESTED** [review]

**Comment by justinfagnani** [comment]
@bicknellr fyi

> Really odd edge case. I guess the code where you found this must have been intentionally using a function in this way, since the problem is that capture isn't set properly in this case?

Yeah, in lit-html we pass the listener from event bindings as both the listener and options arguments. This lets us very easily support an object with `handleEvent` and options, or a method with the `@eventOptions()` decorator - which just sets the options properties on the method.

File: packages/shadydom/src/patch-events.js [hunk 3]
```diff
@@ -44,7 +44,10 @@ const supportsEventOptions = (() => {
 
 const parseEventOptions = (optionsOrCapture) => {
   let capture, once, passive, shadyTarget;
-  if (optionsOrCapture && optionsOrCapture instanceof Object) {
+  if (
+    optionsOrCapture &&
+    (typeof optionsOrCapture === 'object' || optionsOrCapture instanceof Object)
```
Comment (justinfagnani) [inline packages/shadydom/src/patch-events.js:49]: The only values I know of that are not `typeof o === 'object'` but are `o instanceof Object` are functions. `typeof o === 'function'` should always be faster.

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts [hunk 4]
```diff
@@ -38,20 +38,22 @@ if (
...
-    if (optionsOrCapture && optionsOrCapture instanceof Object) {
-      capture = Boolean(optionsOrCapture.capture);
-      once = Boolean(optionsOrCapture.once);
-      passive = Boolean(optionsOrCapture.passive);
+    let capture, once;
+    if (
+      optionsOrCapture &&
+      (typeof optionsOrCapture === 'object' ||
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (optionsOrCapture as any) instanceof Object)
```
Comment (justinfagnani) [inline packages/webcomponentsjs/ts_src/platform/event-listener-options.ts:46]: same here

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    const {capture, once} = parseEventOptions(options);
+    const map = getListenerMap(this, type, capture);
+    let wrapper = map.get(listener);
+    if (wrapper === undefined) {
+      wrapper = (e: Event) => {
+        map.delete(listener);
+        // Try to remove both listener and wrapper since `once` state
+        // is not used as a unique key and we don't track which was added.
+        origRemoveEventListener.call(this, type, wrapper, capture);
+        origRemoveEventListener.call(this, type, listener, capture);
```
Comment (bicknellr) [inline packages/webcomponentsjs/ts_src/platform/event-listener-options.ts:101]: Isn't `wrapper` always the right choice here, since this is happening inside `wrapper`?

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        ((listener as EventListenerObject).handleEvent ?? listener)(e);
+      };
+      map.set(listener, wrapper);
+      const nativeListener = once ? wrapper : listener;
```
Comment (bicknellr) [inline packages/webcomponentsjs/ts_src/platform/event-listener-options.ts:105]: If you added `nativeListener` to the map instead of `wrapper`, I think you could avoid calling `origRemoveEventListener` twice in remove also, since the correct listener would always be whatever comes out of the map.

**Review by bicknellr — CHANGES_REQUESTED** [review]

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    options?: boolean | AddEventListenerOptions | undefined
+  ) {
+    if (listener == null) {
+      return;
+    }
+    const {capture, once} = parseEventOptions(options);
+    const map = getListenerMap(this, type, capture);
+    let wrapper = map.get(listener);
+    if (wrapper === undefined) {
+      wrapper = (e: Event) => {
```
Comment (bicknellr) [inline packages/webcomponentsjs/ts_src/platform/event-listener-options.ts:96]: I think this is the same
```suggestion
    if (!map.has(listener)) {
      const wrapper = (e: Event) => {
```

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts (hunk 5 above)
Comment (sorvell) [reply packages/webcomponentsjs/ts_src/platform/event-listener-options.ts]: Changed based on next suggestion.

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts (hunk 7 above)
Comment (sorvell) [reply packages/webcomponentsjs/ts_src/platform/event-listener-options.ts]: Done.

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts (hunk 6 above)
Comment (sorvell) [reply packages/webcomponentsjs/ts_src/platform/event-listener-options.ts]: Done.

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+      return;
+    }
+    const {capture, once} = parseEventOptions(options);
+    const map = getListenerMap(this, type, capture);
+    if (!map.has(listener)) {
+      const cachedListener = once
+        ? (e: Event) => {
+            map.delete(listener);
+            origRemoveEventListener.call(this, type, cachedListener, capture);
+            ((listener as EventListenerObject).handleEvent ?? listener)(e);
```
Comment (bicknellr) [inline packages/webcomponentsjs/ts_src/platform/event-listener-options.ts:99]: I think this needs to be called with `call` to set up the correct `this`: `this`, when `listener` is the function being called, and `e.currentTarget`, when calling `listener.handleEvent`: if I'm reading the spec correctly, the DOM normally calls `handleEvent` as part of ['inner invoke'](https://dom.spec.whatwg.org/#concept-event-listener-inner-invoke), which calls ['call a user object's operation'](https://heycam.github.io/webidl/#call-a-user-objects-operation) by passing the event's `currentTarget` as `thisArg`. 'Call a user object's operation' also first checks if the callback object it's given is itself callable (calling that, if so) before falling back to trying to read `opName` (`"handleEvent"`).

File: packages/shadydom/src/patch-events.js (hunk 3 above)
Comment (sorvell) [reply packages/shadydom/src/patch-events.js]: Done.

File: packages/shadydom/src/patch-events.js [hunk 9]
```diff
@@ -43,22 +43,23 @@ const supportsEventOptions = (() => {
 })();
 
 const parseEventOptions = (optionsOrCapture) => {
-  let capture, once, passive, shadyTarget;
-  if (optionsOrCapture && typeof optionsOrCapture === 'object') {
+  let capture, once, shadyTarget;
+  if (
+    optionsOrCapture &&
```
Comment (justinfagnani) [inline packages/shadydom/src/patch-events.js:48]: super nit: this only value that `optionsOrCapture &&` is guarding against now is `null`. You could do an explicit check:
```suggestion
    optionsOrCapture !== null &&
```

File: packages/shadydom/src/patch-events.js [hunk 10]
```diff
@@ -379,35 +403,14 @@ function shadyDispatchEvent(e) {
...
-  return (
-    node === savedNode &&
-    type === savedType &&
-    capture === savedCapture &&
-    once === savedOnce &&
-    passive === savedPassive
-  );
+function listenerSettingsEqual(savedListener, node, type, capture) {
+  let {node: savedNode, type: savedType, capture: savedCapture} = savedListener;
+  return node === savedNode && type === savedType && capture === savedCapture;
```
Comment (justinfagnani) [inline packages/shadydom/src/patch-events.js:408]: This is probably ok, but ideally I'd want to check what the browser considers equal for options. ie, should these be loose checks?

File: packages/shadydom/src/patch-events.js [hunk 11]
```diff
@@ -271,6 +272,29 @@ let EventPatches = {
   },
 };
 
+const hasDescriptors = utils.settings.hasDescriptors;
+
+const eventPhaseDescriptor =
+  hasDescriptors &&
+  Object.getOwnPropertyDescriptor(Event.prototype, 'eventPhase');
+if (eventPhaseDescriptor) {
```
Comment (justinfagnani) [inline packages/shadydom/src/patch-events.js:280]: Why is this conditional? It's supported back to IE9. Add a comment?

File: packages/shadydom/src/patch-events.js [hunk 12]
```diff
@@ -271,6 +272,29 @@ let EventPatches = {
   },
 };
 
+const hasDescriptors = utils.settings.hasDescriptors;
```
Comment (justinfagnani) [inline packages/shadydom/src/patch-events.js:275]: Is this related to listener options?

**Review by justinfagnani — COMMENTED** [review]
Should the eventPhase stuff be a separate PR? What was the bug here?

File: packages/shadydom/src/patch-events.js (hunk 10 above)
Comment (sorvell) [reply packages/shadydom/src/patch-events.js:410]: Think these have to exactly match, see https://dom.spec.whatwg.org/#add-an-event-listener.

File: packages/shadydom/src/patch-events.js (hunk 11 above)
Comment (sorvell) [reply packages/shadydom/src/patch-events.js:282]: Added comment. Chrome 41 (a test browser) is missing it.

File: packages/shadydom/src/patch-events.js (hunk 12 above)
Comment (sorvell) [reply packages/shadydom/src/patch-events.js]: Safari 9 has broken DOM descriptors and this catches that. Added a comment.

File: packages/shadydom/src/patch-events.js (hunk 11 above)
Comment (justinfagnani) [reply packages/shadydom/src/patch-events.js:282]: Are we still testing on 41?

Which supported browser does that stand in for now... older Cobalt?

Comment (sorvell) [reply packages/shadydom/src/patch-events.js:282]: Actually no, but it's probably good to continue to test on given the range of browsers needed... internally.

**Comment by bicknellr** [comment]
+1 on splitting the event phase part into a different PR, if possible.

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+        ? (e: Event) => {
+            map.delete(listener);
+            origRemoveEventListener.call(this, type, cachedListener, capture);
+            ((listener as EventListenerObject).handleEvent ?? listener).call(
+              this,
+              e
+            );
+          }
+        : listener;
+      map.set(listener, cachedListener);
```
Comment (justinfagnani) [inline packages/webcomponentsjs/ts_src/platform/event-listener-options.ts:121]: I think if you add the listener as both the key and value it'll be held on to strongly. I would add another value like `undefined`

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+    const map = getListenerMap(this, type, capture);
+    if (!map.has(listener)) {
+      const cachedListener = once
+        ? (e: Event) => {
+            map.delete(listener);
+            origRemoveEventListener.call(this, type, cachedListener, capture);
+            ((listener as EventListenerObject).handleEvent ?? listener).call(
+              this,
+              e
+            );
```
Comment (bicknellr) [inline packages/webcomponentsjs/ts_src/platform/event-listener-options.ts:118]: This first needs to check whether or not `listener` is callable before trying `listener.handleEvent`:

```suggestion
            if (typeof listener === 'function') {
              return listener.call(this, e);
            }

            const handleEvent = listener?.handleEvent;
            if (typeof handleEvent === 'function') {
              return handleEvent.call(e.currentTarget, e);
            }
```

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts (hunk 13 above)
Comment (sorvell) [reply packages/webcomponentsjs/ts_src/platform/event-listener-options.ts:132]: Done.

File: packages/webcomponentsjs/ts_src/platform/event-listener-options.ts [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+      capture,
+      once,
+    };
+  };
+
+  const origAddEventListener = nativeEventTarget.prototype.addEventListener;
+  const origRemoveEventListener =
+    nativeEventTarget.prototype.removeEventListener;
+
+  const captureListenerMap = new WeakMap();
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/webcomponentsjs/ts_src/platform/event-listener-options.ts]: Added types. Don't want to change the project options, but we can make an issue if it's important.

File: packages/tests/webcomponentsjs_/event-listener-options.html [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+            });
+            el1.click();
+            assert.isTrue(
+              functionCalled,
+              'function called when handleEvent is defined on it'
+            );
+            assert.isFalse(
+              handleEventCalled,
+              'handleEvent not called when on listener function'
+            );
```
Comment (bicknellr) [inline packages/tests/webcomponentsjs_/event-listener-options.html:251]: :+1:

File: packages/tests/webcomponentsjs_/event-listener-options.html [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+            el1.dispatchEvent(new Event('click'));
+            assert.equal(callback.callCount, 0, 'not called if removed early');
+            el1.addEventListener('click', callback, {once: true});
+            el1.removeEventListener('click', callback, {once: true});
+            el2.addEventListener('click', callback, {once: true});
+            el1.dispatchEvent(new Event('click'));
+            assert.equal(callback.callCount, 0, 'not called if removed early');
+            el2.dispatchEvent(new Event('click'));
+            el2.dispatchEvent(new Event('click'));
+            assert.equal(callback.callCount, 1, 'not called if removed early');
```
Comment (bicknellr) [inline packages/tests/webcomponentsjs_/event-listener-options.html:300]: The assert message feels off for this case.

File: packages/tests/webcomponentsjs_/event-listener-options.html [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+              cCurrentTarget = e.currentTarget;
+            };
+            parent.addEventListener('test', pListener, {once: true});
+            child.addEventListener('test', cListener, {once: true});
+            gchild.dispatchEvent(new Event('test', {bubbles: true}));
+            assert.ok(pThisRef);
+            assert.equal(pThisRef, pCurrentTarget);
+            assert.ok(cThisRef);
+            assert.equal(cThisRef, cCurrentTarget);
+          });
```
Comment (bicknellr) [inline packages/tests/webcomponentsjs_/event-listener-options.html:360]: Could you add one for the `handleEvent` case also?

File: packages/shadydom/src/patch-events.js [hunk 19]
```diff
@@ -271,6 +272,31 @@ let EventPatches = {
   },
 };
 
+// Some browsers have missing or broken descriptors. If this is the case
+// eventPhase can't be polyfilled. For example, Chrome 41 does not
+// have prototypical descriptors on DOM objects and Safari 9 has broken ones
+// that cannot be used.
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/shadydom/src/patch-events.js:278]: Don't think it's worth it since this affects only these really old browsers and `eventPhase` generally isn't critically important. We'd likely need to do the manual `dispatchEvent` always on these browsers to fix it, and that doesn't seem worth it.

File: packages/shadydom/src/patch-events.js [hunk 20]
```diff
@@ -271,6 +272,31 @@ let EventPatches = {
...
+const eventPhaseDescriptor =
+  utils.settings.hasDescriptors &&
+  Object.getOwnPropertyDescriptor(Event.prototype, 'eventPhase');
+if (eventPhaseDescriptor) {
+  Object.defineProperty(EventPatches, 'eventPhase', {
+    get() {
+      return this.currentTarget === this.target
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/shadydom/src/patch-events.js:285]: Events currently aren't patched to have `SHADY_PREFIX` properties. The patches for elements and events are treated differently.

Element prototypes are patched and in noPatch mode, they aren't patched. We access the `SHADY_PREFIX` properties because they always exist and if a noPatch wrapper calls a method, we can safely call the `SHADY_PREFIX` versions.

Event instances are patched so we will only ever get to this code on a patched instance. I can't say for sure why the Event prototype isn't patched and it's something perhaps we can revisit, but this is how it works right now.

File: packages/tests/shadydom/shady.html [hunk 21]
```diff
@@ -1301,11 +1383,91 @@
...
+          );
+          assert.equal(callback.callCount, 2, 'listener not called');
+        });
+
         test('support `once` option', function () {
           ShadyDOM.wrap(el1).addEventListener('click', callback, {once: true});
-          el1.click();
+          ShadyDOM.wrapIfNeeded(el1).dispatchEvent(
+            new Event('click', {bubbles: true, composed: true})
+          );
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/tests/shadydom/shady.html:1466]: We need to ensure the event dispatches through ShadyDOM. In `noPatch` mode it will not. In general we use `wrap` to ensure this, but the wrapper does not currently have `click` (a separate issue).

File: packages/tests/webcomponentsjs_/event-listener-options.html (hunk 17 above)
Comment (sorvell) [reply packages/tests/webcomponentsjs_/event-listener-options.html:300]: Fixed assert messages.

File: packages/tests/shadydom/shady.html [hunk 22]
```diff
@@ -1463,6 +1659,70 @@
...
+          ShadyDOM.wrapIfNeeded(h1_c_h2_c_h3_c).addEventListener(
+            'click',
+            listener
+          );
+          ShadyDOM.wrapIfNeeded(h1_c_h2_c_h3_c).dispatchEvent(
+            new Event('click', {bubbles: true, composed: true})
+          );
+          const t = Event.AT_TARGET;
+          const b = Event.BUBBLING_PHASE;
+          assert.deepEqual(eventPhases, [t, t, b, t, b, t]);
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/tests/shadydom/shady.html:1687]: It does. Added this into the tests.

File: packages/shadydom/src/patch-events.js [hunk 23]
```diff
@@ -514,7 +515,7 @@ export function addEventListener(type, fnOrObj, optionsOrCapture) {
    * @this {HTMLElement}
    * @param {Event} e
    */
-  const wrapperFn = function (e) {
+  const wrapperFn = (e) => {
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/shadydom/src/patch-events.js:518]: Think this was just a mistake. Reverting, but the @this here **should** be EventTarget so will update that.

File: packages/tests/webcomponentsjs_/event-listener-options.html (hunk 18 above)
Comment (sorvell) [reply packages/tests/webcomponentsjs_/event-listener-options.html:360]: Added and this revealed an issue. I can't find it in the spec, but via empirical testing, `this` is the listener object itself when listening with an object with `handleEvent`.

Comment (bicknellr) [reply packages/tests/webcomponentsjs_/event-listener-options.html:365]: Ok, if this is what browsers do then that sounds good.

File: packages/webcomponentsjs/CHANGELOG.md [hunk 24]
```diff
@@ -8,6 +8,12 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 <!-- ## [Unreleased] -->
 
+## [Unreleased]
+
+- Polyfill `addEventListener/removeEventListener` event listener options,
+  including `{capture: boolean, once: boolean}`
+  ([#469]https://github.com/webcomponents/polyfills/pull/469)
```
[the remark this answers is not in the document]
Comment (bicknellr) [reply packages/webcomponentsjs/CHANGELOG.md:15]: just want to make sure you saw this one

File: packages/tests/webcomponentsjs_/event-listener-options.html (hunk 18 above)
Comment (bicknellr) [reply packages/tests/webcomponentsjs_/event-listener-options.html:365]: I misread the spec. From [call a user object's operation](https://heycam.github.io/webidl/#call-a-user-objects-operation), step 10.5:

> Set *thisArg* to *O* (overriding the provided value).

Where *O* is the listener object at that point.

File: packages/tests/shadydom/shady.html (hunk 21 above)
[the remark this answers is not in the document]
Comment (bicknellr) [reply packages/tests/shadydom/shady.html:1466]: Isn't `once` handled by `addEventListener` and the listener itself? I'm confused about which part a browser-dispatched event would miss.

File: packages/webcomponentsjs/CHANGELOG.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/webcomponentsjs/CHANGELOG.md:15]: Looks good.

File: packages/tests/shadydom/shady.html (hunk 21 above)
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/tests/shadydom/shady.html:1466]: If the event is not dispatched via ShadyDOM, it won't get the `composed` flag on browsers that don't support it. We depend on that flag in the listeners.

File: packages/webcomponentsjs/CHANGELOG.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (bicknellr) [reply packages/webcomponentsjs/CHANGELOG.md:15]: It needs those extra parens though.

File: packages/tests/shadydom/shady.html (hunk 21 above)
[the remark this answers is not in the document]
Comment (bicknellr) [reply packages/tests/shadydom/shady.html:1466]: Oh, now I get it, so this change isn't specifically related to this test but rather for the others where there are polyfilled shadow roots?

**Review by bicknellr — APPROVED** [review]
Other than the parens in the readme, LGTM
